feat: MathML support

sto3psl opened the pull request:
Implements support for [MathML](https://developer.mozilla.org/en-US/docs/Web/MathML) elements in Vue.

> Mathematical Markup Language (MathML) is an [XML](https://developer.mozilla.org/en-US/docs/Web/XML)-based language for describing mathematical notation.

Similar to SVG, MathML requires a different namespace when creating elements in the browser. My changes update the widely used `isSVG` bool into an enum, supporting SVG, MathML and of course HTML. I mostly copied the SVG related code and adapted it for MathML.

This is the first time I'm contributing to a larger project. I'd be very happy about any feedback and suggestions!

Related Issue: https://github.com/vuejs/core/issues/7820

sxzz commented on packages/runtime-dom/src/index.ts:
```diff
@@ -99,7 +100,10 @@ export const createApp = ((...args) => {
-    const proxy = mount(container, false, container instanceof SVGElement)
+    let namespace: 'svg' | 'mathml' | undefined
+    if (container instanceof SVGElement) namespace = 'svg'
+    if (container instanceof MathMLElement) namespace = 'mathml'
```
```suggestion
    else if (container instanceof MathMLElement) namespace = 'mathml'
```

sxzz commented on packages/runtime-dom/src/index.ts:
```diff
@@ -122,7 +126,10 @@ export const createSSRApp = ((...args) => {
-      return mount(container, true, container instanceof SVGElement)
+      let namespace: 'svg' | 'mathml' | undefined
+      if (container instanceof SVGElement) namespace = 'svg'
+      if (container instanceof MathMLElement) namespace = 'mathml'
```
```suggestion
      else if (container instanceof MathMLElement) namespace = 'mathml'
```

sto3psl commented:
> It's a breaking change to Vue, right? Can we avoid it?
> 
> Maybe we can change API `isSVG?: boolean` to `namespace?: boolean | 'svg' | 'mathml'`. If `namespace` is `true` then fallback to `svg`.

I don't think so, since this is only internal code. As far as I can tell, no public APIs have changed. Please correct me if I'm wrong.

sxzz commented on packages/runtime-core/src/renderer.ts:
```diff
@@ -580,19 +580,25 @@ function baseCreateRenderer(
     optimized: boolean
   ) => {
-    isSVG = isSVG || (n2.type as string) === 'svg'
+    if ((n2.type as string) === 'svg') {
+      namespace = 'svg'
+    }
+    if ((n2.type as string) === 'math') {
+      namespace = 'mathml'
+    }
```
Prefer `switch` here I think

sxzz commented on packages/runtime-core/src/renderer.ts:
```diff
@@ -794,7 +809,7 @@ function baseCreateRenderer(
     parentComponent: ComponentInternalInstance | null,
     parentSuspense: SuspenseBoundary | null,
-    isSVG: boolean,
+    namespace: 'svg' | 'mathml' | undefined,
```
Maybe we can extract the type of namespace instead of repeating code.

sto3psl replied:
Sure! Where should I put it? It's used in quite a lot of files.

sxzz commented on packages/runtime-core/src/compat/global.ts:
```diff
@@ -503,15 +503,17 @@ function installCompatMount(
-      const isSVG = container instanceof SVGElement
+      let namespace: 'svg' | 'mathml' | undefined
+      if (container instanceof SVGElement) namespace = 'svg'
+      if (container instanceof MathMLElement) namespace = 'mathml'
```
```suggestion
      else if (container instanceof MathMLElement) namespace = 'mathml'
```

sxzz commented:
/ecosystem-ci run

sxzz commented:
> I don't think so, since this is only internal code. As far as I can tell, no public APIs have changed. Please correct me if I'm wrong.

<img width="668" alt="image" src="https://user-images.githubusercontent.com/6481596/223122726-02503edd-e1f1-46ec-89d1-c5044bbf045a.png">
It's a public API definitely.

sto3psl commented:
> > I don't think so, since this is only internal code. As far as I can tell, no public APIs have changed. Please correct me if I'm wrong.
> 
> <img alt="image" width="668" src="https://user-images.githubusercontent.com/6481596/223122726-02503edd-e1f1-46ec-89d1-c5044bbf045a.png">
> 
> It's a public API definitely.

I just saw that, too! Thanks for checking. Will change it to `boolean | 'svg' | 'mathml' | undefined`.

sto3psl commented:
Not sure why `ci / unit-test` fails. It works fine for me locally and the tests pass.

sxzz commented:
I pushed a commit to fix unit tests. Please don't forget to pull before committing

sto3psl commented:
@sxzz I implemented the backwards compat and reused the type everywhere. Added a comment too that will show up in Intellisense if one wonders why that union type is a bit weird 😄.

<img width="648" alt="Display union type with backwards compatibility" src="https://user-images.githubusercontent.com/5798652/223203416-ae16f91d-7071-44ec-8803-d1fe050eea3d.png">

sto3psl commented:
I checked why the ecosystem-ci fails for a lot of suits and it's the same reason a lot of core tests failed here first.

> **Warning**
> `MathMLElement is not defined`

It looks like JSDOM doesn't provide `MathMLElement` as global. I worked around it by stubbing it with Vitest. Is there anything I can do about this?

sxzz commented:
MathMLElement API is [introduced in Chrome > 109](https://caniuse.com/mdn-api_mathmlelement), so we should do some checks to ensure it's compatible with all ES6 browsers. `typeof MathMLElement === 'function' && el instanceof MathMLElement`

sxzz commented on packages/runtime-core/src/renderer.ts:
```diff
@@ -82,10 +82,15 @@ export interface HydrationRenderer extends Renderer<Element | ShadowRoot> {
+/**
+ * `boolean` is kept for backwards compatibility with the old `isSVG` attribute.
+ */
+export type ElementNamespace = boolean | 'svg' | 'mathml' | undefined
```
```suggestion
export type ElementNamespace = 'svg' | 'mathml' | undefined
```

sxzz commented on packages/runtime-core/src/apiCreateApp.ts:
```diff
@@ -46,7 +46,7 @@ export interface App<HostElement = any> {
     rootContainer: HostElement | string,
     isHydrate?: boolean,
-    isSVG?: boolean
+    namespace?: ElementNamespace
```
We should add the boolean type here for public API only and normalize the namespace (convert boolean to string) and pass the normalized value for internal API.

sto3psl commented:
@sxzz I added the guard for missing `MathMLElement`s and normalized the type in `app.mount`. I didn't find other public APIs, do you know of any? After checking the Vue docs, the 2nd and 3rd argument to [`app.mount`](https://vuejs.org/api/application.html#app-mount) aren't even documented (which is fine in my opinion).

sxzz commented on packages/runtime-core/src/components/Teleport.ts:
```diff
@@ -108,7 +112,12 @@ export const TeleportImpl = {
+        if (namespace === 'svg' || isTargetSVG(target)) {
+          namespace = 'svg'
+        }
+        if (namespace === 'mathml' || isTargetMathML(target)) {
```
```suggestion
        else if (namespace === 'mathml' || isTargetMathML(target)) {
```

sxzz commented on packages/runtime-core/src/components/Teleport.ts:
```diff
@@ -144,7 +153,13 @@ export const TeleportImpl = {
+      if (namespace === 'svg' || isTargetSVG(target)) {
+        namespace = 'svg'
+      }
+      if (namespace === 'mathml' || isTargetMathML(target)) {
```
```suggestion
      else if (namespace === 'mathml' || isTargetMathML(target)) {
```

sto3psl commented:
Seems like Nuxt doesn't like the bundle size increase and naive-ui seems unrelated(?)

sto3psl commented:
@sxzz @sodatea is there a chance that this change might land in Vue 3.3?

sxzz commented:
@sto3psl Yes I think so!

sto3psl commented:
Since 3.3 is in beta now it would be amazing if this feature could land as well! What do you think @yyx990803?

sto3psl commented:
Hey! Is there anything I can do to  get this merged?

It would be so nice to have proper MathML support in a future Vue version.

sxzz reviewed (approved):
I hope this can be included in the next minor release, version 3.4.

50Wliu commented on packages/runtime-core/src/renderer.ts:
```diff
@@ -82,10 +82,15 @@ export interface HydrationRenderer extends Renderer<Element | ShadowRoot> {
 }
 
+/**
+ * `boolean` is kept for backwards compatibility with the old `isSVG` attribute.
```
This comment doesn't seem accurate anymore :)

sto3psl replied:
You're right. Removed it.

50Wliu reviewed (commented):
Was also looking to add MathML to my Vue app today and was disappointed when it didn't render correctly, but glad to see that it's in the works :).
I see this PR has been approved; is there anything else that needs to be done for it to be merged?

xarthurx commented:
Any updates on this PR? Math support is really needed.

sto3psl commented:
> Any updates on this PR? Math support is really needed.

My hope is that it lands whenever Vue 3.4 is released. No idea when that's planned but there is an open PR which is referenced above.

sxzz commented:
See https://github.com/vuejs/core/pull/9249#issuecomment-1772442069

I think we should also treat MathML tags as "late added tags"

sto3psl commented:
@sxzz sure I can do that once the mentioned PR has landed. 

Is there any plan to reset the new tag list with am major update in the future?

I assume that these won't stay new forever.

Do people that have vue components with reserved names get a runtime warning?
